### PR TITLE
Catch error arising in `as_variant_sub` when scrutinee is not variant

### LINKED ISSUE #2934: Pattern inference crash when variant `switch` alternative contradicts scrutinee's type
Stderr:
OOPS! You've triggered a compiler bug.
Please report this at https://github.com/dfinity/motoko/issues/new with the following details:

Motoko 0.6.14 (source i5rk4rv7-xsrcz16w-6xp3834q-icgfd9zl)

Fatal error: exception (Invalid_argument Type.as_variant_sub)
Raised at Mo_types__Type.invalid in file "mo_types/type.ml", line 357, characters 16-54
Called from Mo_frontend__Typing.check_pat' in file "mo_frontend/typing.ml", line 1804, characters
 41-67
Called from Mo_frontend__Typing.check_pat in file "mo_frontend/typing.ml", line 1747, characters 
11-32
Called from Mo_frontend__Typing.check_pat in file "mo_frontend/typing.ml", line 1747, characters 
11-32
Called from Mo_frontend__Typing.check_case in file "mo_frontend/typing.ml", line 1630, characters
 11-34
Called from Stdlib__list.iter in file "list.ml", line 110, characters 12-15
Called from Mo_frontend__Typing.check_exp' in file "mo_frontend/typing.ml", line 1451, characters
 4-30
Called from Mo_frontend__Typing.check_exp in file "mo_frontend/typing.ml", line 1351, characters 
11-45
Called from Mo_frontend__Typing.check_dec in file "mo_frontend/typing.ml", line 2223, characters 
4-23
Called from Mo_frontend__Typing.recover_with in file "mo_frontend/typing.ml", line 77, characters
 56-59
Called from Mo_frontend__Typing.recover in file "mo_frontend/typing.ml" (inlined), line 79, chara
cters 18-37
Called from Mo_frontend__Typing.check_block_exps in file "mo_frontend/typing.ml", line 2217, char
acters 4-38
Called from Mo_frontend__Typing.recover_with in file "mo_frontend/typing.ml", line 77, characters
 56-59
Called from Mo_frontend__Typing.check_block in file "mo_frontend/typing.ml", line 2205, character
s 2-47
Called from Mo_frontend__Typing.check_exp' in file "mo_frontend/typing.ml", line 1442, characters
 11-42
Called from Mo_frontend__Typing.check_exp in file "mo_frontend/typing.ml", line 1351, characters 
11-45
Called from Mo_frontend__Typing.infer_exp'' in file "mo_frontend/typing.ml", line 1119, character
s 6-57
Called from Mo_frontend__Typing.infer_exp' in file "mo_frontend/typing.ml", line 885, characters 
10-29
Called from Mo_frontend__Typing.infer_exp in file "mo_frontend/typing.ml" (inlined), line 869, ch
aracters 2-31
Called from Mo_frontend__Typing.infer_dec in file "mo_frontend/typing.ml", line 2147, characters 
4-21
Called from Mo_frontend__Typing.check_dec in file "mo_frontend/typing.ml", line 2226, characters 
13-30
Called from Mo_frontend__Typing.recover_with in file "mo_frontend/typing.ml", line 77, characters
 56-59
Called from Mo_frontend__Typing.recover in file "mo_frontend/typing.ml" (inlined), line 79, chara
cters 18-37
Called from Mo_frontend__Typing.infer_block_exps in file "mo_frontend/typing.ml", line 2139, char
acters 24-58
Called from Mo_frontend__Typing.infer_block in file "mo_frontend/typing.ml", line 2120, character
s 10-55
Called from Mo_frontend__Typing.infer_obj in file "mo_frontend/typing.ml", line 2002, characters 
17-40
Called from Mo_frontend__Typing.infer_exp' in file "mo_frontend/typing.ml", line 885, characters 
10-29
Called from Mo_frontend__Typing.infer_exp in file "mo_frontend/typing.ml" (inlined), line 869, ch
aracters 2-31
Called from Mo_frontend__Typing.infer_dec in file "mo_frontend/typing.ml", line 2147, characters 
4-21
Called from Mo_frontend__Typing.infer_block in file "mo_frontend/typing.ml", line 2120, character
s 10-55
Called from Mo_frontend__Typing.check_lib.(fun) in file "mo_frontend/typing.ml", line 2563, chara
cters 23-59
Called from Mo_frontend__Typing.recover_opt.(fun) in file "mo_frontend/typing.ml", line 78, chara
cters 55-60
Called from Mo_frontend__Typing.recover_with in file "mo_frontend/typing.ml", line 77, characters
 56-59
Called from Diag.with_message_store in file "lang_utils/diag.ml", line 82, characters 10-13
Called from Pipeline.check_lib in file "pipeline/pipeline.ml", line 200, characters 16-41
Called from Pipeline.chase_imports.go in file "pipeline/pipeline.ml", line 362, characters 22-41
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.traverse_ in file "lang_utils/diag.ml", line 48, characters 20-25
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Pipeline.chase_imports.go in file "pipeline/pipeline.ml", line 360, characters 18-37
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.traverse_ in file "lang_utils/diag.ml", line 48, characters 20-25
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Pipeline.chase_imports.go_set in file "pipeline/pipeline.ml" (inlined), line 391, cha
racters 20-42
Called from Pipeline.chase_imports in file "pipeline/pipeline.ml", line 393, characters 47-63
Called from Pipeline.load_progs in file "pipeline/pipeline.ml", line 401, characters 21-52
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Pipeline.compile_files in file "pipeline/pipeline.ml", line 678, characters 27-71
Called from Moc.process_files in file "exes/moc.ml", line 188, characters 49-94
Called from Moc in file "exes/moc.ml", line 282, characters 4-23

### PR DESCRIPTION
**Pull request by ggreif**
Fixes #2934.

### CODE REVIEW

File: src/mo_frontend/typing.ml [hunk 1]
```diff
@@ -1800,10 +1800,13 @@ and check_pat' env t pat : Scope.val_env =
         display_typ_expand t
     in check_pat env t1 pat1
   | TagP (id, pat1) ->
-    let t1 =
-      match T.lookup_val_field_opt id.it (T.as_variant_sub id.it t) with
-      | Some t1 -> t1
-      | None -> T.Non
+    let t1 = try
+        match T.lookup_val_field_opt id.it (T.as_variant_sub id.it t) with
```
Comment (crusso) [inline src/mo_frontend/typing.ml:1804]:
```suggestion
     match T.lookup_val_field_opt id.it (T.as_variant_sub id.it t) with
```

**Review by crusso — APPROVED** [review]
LGTM

File: src/mo_frontend/typing.ml (hunk 1 above)
Comment (ggreif) [reply src/mo_frontend/typing.ml:1805]: indenting by 1 seemed strange, how about this?

**Comment by rossberg** [comment]
Thanks Gabor!

**Pull request merged**